### PR TITLE
Pubkey converter refactor

### PR DESCRIPTION
**Pull request by schimih**
- Changed `PubkeyConverter` interface:
  - `Encode()` now returns also an error;
  -  Moved `EncodeBytesSlice()` from `/elrond-vm-common` repo and added it as `EncodeSlice()`;

- Replaced the logger from the `PubkeyConverter` constructor with a prefix. This prefix will be provided as config variable from `config.toml`.

- Added `EncodeSlice()` also to the hexPubkeyConverter in order to respect the `PubkeyConverter` even though this Encoding will never return an error;

### CODE REVIEW

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 1]
```diff
@@ -80,42 +77,39 @@ func (bpc *bech32PubkeyConverter) Decode(humanReadable string) ([]byte, error) {
...
 	conv, err := bech32.ConvertBits(pkBytes, bech32Config.fromBits, bech32Config.toBits, bech32Config.pad)
 	if err != nil {
-		bpc.log.Warn("bech32PubkeyConverter.Encode ConvertBits",
-			"hex buff", hex.EncodeToString(pkBytes),
-			"error", err,
-			"stack trace", string(debug.Stack()),
-		)
-
-		return ""
+		return "", fmt.Errorf("%w when converting bits", ErrConvertBits)
```
Comment (sstanculeanu) [inline core/pubkeyConverter/bech32PubkeyConverter.go:89]: according to:
`var ErrConvertBits = errors.New("invalid fromBits or toBits when converting bits")`

this will return `invalid fromBits or toBits when converting bits when converting bits`

why not use `ErrBech32ConvertError` as on L68?

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 2]
```diff
@@ -80,42 +77,39 @@ func (bpc *bech32PubkeyConverter) Decode(humanReadable string) ([]byte, error) {
...
-		)
-
-		return ""
+		return "", fmt.Errorf("%w when encoding address", ErrConvertBits)
+	}
+
+	return converted, nil
+}
+
+// EncodeSlice will encode the provided bytes slice
```
Comment (sstanculeanu) [inline core/pubkeyConverter/bech32PubkeyConverter.go:100]:
```suggestion
// EncodeSlice converts the provided bytes slice into a slice of bech32 addresses
```
as to keep same format as for `Encode`

File: core/pubkeyConverter/errors.go [hunk 3]
```diff
@@ -13,3 +13,9 @@ var ErrInvalidErdAddress = errors.New("invalid ERD address")
 
 //ErrBech32ConvertError signals that conversion the 5bit alphabet to 8bit failed
 var ErrBech32ConvertError = errors.New("can't convert bech32 string")
+
+//ErrEmptyPrefix signals that an empty prefix has been used
```
Comment (sstanculeanu) [inline core/pubkeyConverter/errors.go:17]:
```suggestion
// ErrEmptyPrefix signals that an empty prefix has been used
```

File: data/mock/pubkeyConverterStub.go [hunk 4]
```diff
@@ -26,12 +26,12 @@ func (pcs *PubkeyConverterStub) Decode(humanReadable string) ([]byte, error) {
 }
 
 // Encode -
-func (pcs *PubkeyConverterStub) Encode(pkBytes []byte) string {
+func (pcs *PubkeyConverterStub) Encode(pkBytes []byte) (string, error) {
 	if pcs.EncodeCalled != nil {
-		return pcs.EncodeCalled(pkBytes)
+		return pcs.EncodeCalled(pkBytes), nil
```
Comment (sstanculeanu) [inline data/mock/pubkeyConverterStub.go:31]: L7 should be `Encode(pkBytes []byte) (string, error)`, this way L31 becomes `return pcs.EncodeCalled(pkBytes)`

File: data/transaction/transaction.go [hunk 5]
```diff
@@ -65,11 +65,21 @@ func (tx *Transaction) GetDataForSigning(encoder Encoder, marshalizer Marshalize
 		return nil, ErrNilMarshalizer
 	}
 
+	receiverAddr, err := encoder.Encode(tx.RcvAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	senderAddr, _ := encoder.Encode(tx.SndAddr)
```
Comment (sstanculeanu) [inline data/transaction/transaction.go:73]:
```suggestion
	senderAddr, err := encoder.Encode(tx.SndAddr)
```

File: core/pubkeyConverter/bech32PubkeyConverter.go (hunk 1 above)
Comment (schimih) [reply core/pubkeyConverter/bech32PubkeyConverter.go:89]: wanted to differentiate between the two functions.
Changed L89 + L94

File: data/mock/pubkeyConverterStub.go (hunk 4 above)
Comment (schimih) [reply data/mock/pubkeyConverterStub.go:31]: missed it :facepalm:

**Review by sstanculeanu — DISMISSED** [review]

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 6]
```diff
@@ -80,42 +77,39 @@ func (bpc *bech32PubkeyConverter) Decode(humanReadable string) ([]byte, error) {
...
 	conv, err := bech32.ConvertBits(pkBytes, bech32Config.fromBits, bech32Config.toBits, bech32Config.pad)
 	if err != nil {
-		bpc.log.Warn("bech32PubkeyConverter.Encode ConvertBits",
-			"hex buff", hex.EncodeToString(pkBytes),
-			"error", err,
-			"stack trace", string(debug.Stack()),
-		)
-
-		return ""
+		return "", ErrConvertBits
```
Comment (ssd04) [inline core/pubkeyConverter/bech32PubkeyConverter.go:89]: i suggest to return the `err`, or to at least include it in the returned err

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 7]
```diff
@@ -80,42 +77,39 @@ func (bpc *bech32PubkeyConverter) Decode(humanReadable string) ([]byte, error) {
...
 	if err != nil {
-		bpc.log.Warn("bech32PubkeyConverter.Encode Encode",
-			"hex buff", hex.EncodeToString(pkBytes),
-			"conv", hex.EncodeToString(conv),
-			"error", err,
-			"stack trace", string(debug.Stack()),
-		)
-
-		return ""
+		return "", ErrBech32ConvertError
```
Comment (ssd04) [inline core/pubkeyConverter/bech32PubkeyConverter.go:94]: same here

File: core/pubkeyConverter/bech32PubkeyConverter_test.go [hunk 8]
```diff
@@ -96,17 +97,44 @@ func TestBech32PubkeyConverter_EncodeDecodeShouldWork(t *testing.T) {
...
 	buff = []byte("1234567890123456789012345678901234567890")
-	str = bpc.Encode(buff)
+	str, err = bpc.Encode(buff)
+
+	assert.True(t, errors.Is(err, pubkeyConverter.ErrWrongSize))
 	assert.Equal(t, "", str)
 }
+
+func TestBech32PubkeyConverter_EncodeSliceShouldWork(t *testing.T) {
+	addressLen := 32
```
Comment (ssd04) [inline core/pubkeyConverter/bech32PubkeyConverter_test.go:122]: add `t.Parallel()` maybe also in the other tests from above

File: core/pubkeyConverter/hexPubkeyConverter_test.go [hunk 9]
```diff
@@ -83,7 +83,28 @@ func TestHexPubkeyConverter_EncodeDecodeShouldWork(t *testing.T) {
...
 
-	revoveredValue := hpc.Encode(buff)
+	recoveredValue, _ := hpc.Encode(buff)
 
-	assert.Equal(t, value, revoveredValue)
+	assert.Equal(t, value, recoveredValue)
+}
+
+func TestHexPubkeyConverter_EncodeSliceShouldWork(t *testing.T) {
+	addressLen := 16
```
Comment (ssd04) [inline core/pubkeyConverter/hexPubkeyConverter_test.go:92]: parallel run

File: core/check/ifEmpty.go [hunk 10]
```diff
@@ -0,0 +1,10 @@
+package check
+
+import (
+	"strings"
+)
+
+// IfEmpty tests if the provided string is technically empty
+func IfEmpty(s string) bool {
+	return strings.TrimSpace(s) == ""
```
Comment (iulianpascalau) [inline core/check/ifEmpty.go:9]: is this enough? What happens on this string: "\t "

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 11]
```diff
@@ -18,35 +15,35 @@ type config struct {
...
 			ErrInvalidAddressLength)
 	}
 	if addressLen%2 == 1 {
-		return nil, fmt.Errorf("%w when creating hex address converter, addressLen should have been an even number",
+		return nil, fmt.Errorf("%w when creating address converter, addressLen should have been an even number",
 			ErrInvalidAddressLength)
 	}
-	if check.IfNil(log) {
-		return nil, core.ErrNilLogger
+	if check.IfEmpty(prefix) {
```
Comment (iulianpascalau) [inline core/pubkeyConverter/bech32PubkeyConverter.go:39]: :+1:

File: core/pubkeyConverter/bech32PubkeyConverter.go (hunk 6 above)
Comment (iulianpascalau) [reply core/pubkeyConverter/bech32PubkeyConverter.go:89]: ...or wrap it:
```go
fmt.Errorf("%w: %s", ErrConvertBits, err.Error())
```

File: core/pubkeyConverter/bech32PubkeyConverter.go (hunk 7 above)
Comment (iulianpascalau) [reply core/pubkeyConverter/bech32PubkeyConverter.go:94]: yup, wrapping will work

File: core/pubkeyConverter/bech32PubkeyConverter_test.go [hunk 12]
```diff
@@ -1,29 +1,29 @@
...
-	"github.com/ElrondNetwork/elrond-go-core/core/mock"
 	"github.com/ElrondNetwork/elrond-go-core/core/pubkeyConverter"
 	"github.com/stretchr/testify/assert"
 )
 
 func TestNewBech32PubkeyConverter_InvalidSizeShouldErr(t *testing.T) {
 	t.Parallel()
 
-	bpc, err := pubkeyConverter.NewBech32PubkeyConverter(-1, &mock.LoggerMock{})
+	bpc, err := pubkeyConverter.NewBech32PubkeyConverter(-1, "erd")
```
Comment (iulianpascalau) [inline core/pubkeyConverter/bech32PubkeyConverter_test.go:18]: all these "erd" could have been declared as a const

File: core/pubkeyConverter/hexPubkeyConverter_test.go [hunk 13]
```diff
@@ -83,7 +83,28 @@ func TestHexPubkeyConverter_EncodeDecodeShouldWork(t *testing.T) {
 	buff, err := hpc.Decode(value)
 	assert.Nil(t, err)
 
-	revoveredValue := hpc.Encode(buff)
+	recoveredValue, _ := hpc.Encode(buff)
```
Comment (iulianpascalau) [inline core/pubkeyConverter/hexPubkeyConverter_test.go:86]: please also test that the returned err is nil

File: core/check/ifEmpty.go (hunk 10 above)
Comment (schimih) [reply core/check/ifEmpty.go]: changed with `IfHrp()`

File: core/pubkeyConverter/bech32PubkeyConverter_test.go (hunk 12 above)
Comment (schimih) [reply core/pubkeyConverter/bech32PubkeyConverter_test.go]: const in `core/constants` added.

File: core/pubkeyConverter/hexPubkeyConverter_test.go (hunk 13 above)
Comment (schimih) [reply core/pubkeyConverter/hexPubkeyConverter_test.go]: done

File: data/mock/pubkeyConverterStub.go [hunk 14]
```diff
@@ -26,11 +30,29 @@ func (pcs *PubkeyConverterStub) Decode(humanReadable string) ([]byte, error) {
...
+	if pcs.EncodeSliceCalled != nil {
+		return pcs.EncodeSliceCalled(pkBytesSlice)
+	}
+
+	return make([]string, 0), nil
+}
+
+// SilentEncode -
+func (pcs *PubkeyConverterStub) SilentEncode(pkBytes []byte, log core.Logger) string {
+	if pcs.EncodeCalled != nil {
```
Comment (ssd04) [inline data/mock/pubkeyConverterStub.go:52]:
```suggestion
	if pcs.SilentEncodeCalled != nil {
```

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 15]
```diff
@@ -80,44 +80,72 @@ func (bpc *bech32PubkeyConverter) Decode(humanReadable string) ([]byte, error) {
...
+	if err != nil {
+		return "", fmt.Errorf("%w: %s", ErrBech32ConvertError, err.Error())
+	}
+
+	return converted, nil
+}
+
+// SilentEncode converts the provided bytes in a bech32 form without returning any error
+func (bpc *bech32PubkeyConverter) SilentEncode(pkBytes []byte, log core.Logger) string {
+	if len(pkBytes) != bpc.len {
```
Comment (sstanculeanu) [inline core/pubkeyConverter/bech32PubkeyConverter.go:104]: `Encode` method can be reused here thus reducing this method to a call to Encode + a log.Warn in case of error

File: core/pubkeyConverter/errors.go [hunk 16]
```diff
@@ -11,5 +11,11 @@ var ErrWrongSize = errors.New("wrong size")
 // ErrInvalidErdAddress signals that the provided address is not an ERD address
 var ErrInvalidErdAddress = errors.New("invalid ERD address")
 
-//ErrBech32ConvertError signals that conversion the 5bit alphabet to 8bit failed
+// ErrBech32ConvertError signals that conversion the 5bit alphabet to 8bit failed
 var ErrBech32ConvertError = errors.New("can't convert bech32 string")
+
+// ErrHrpPrefix signals that the prefix is not human readable or empty
+var ErrHrpPrefix = errors.New("introduced prefix which is not human readable")
```
Comment (sstanculeanu) [inline core/pubkeyConverter/errors.go:18]: perhaps
```suggestion
var ErrInvalidHrpPrefix = errors.New("invalid hrp prefix")
```
?

**Review by ssd04 — APPROVED** [review]

**Review by sstanculeanu — DISMISSED** [review]

**Review by ssd04 — DISMISSED** [review]

File: core/pubkeyConverter/bech32PubkeyConverter.go [hunk 17]
```diff
@@ -83,18 +86,31 @@ func (bpc *bech32PubkeyConverter) Encode(pkBytes []byte) (string, error) {
...
-		return "", ErrBech32ConvertError
+		return "", fmt.Errorf("%w: %s", ErrBech32ConvertError, err.Error())
 	}
 
-	return converted, nil
+	return encodedBytes, nil
+}
+
+// SilentEncode converts the provided bytes in a bech32 form without returning any error
+func (bpc *bech32PubkeyConverter) SilentEncode(pkBytes []byte, log core.Logger) string {
```
Comment (iulianpascalau) [inline core/pubkeyConverter/bech32PubkeyConverter.go:103]: a unit test for this?

File: core/pubkeyConverter/hexPubkeyConverter.go [hunk 18]
```diff
@@ -57,6 +59,11 @@ func (ppc *hexPubkeyConverter) EncodeSlice(pkBytesSlice [][]byte) ([]string, err
 	return encodedSlice, nil
 }
 
+// SilentEncode converts the provided bytes in a form that this converter can decode. In this case it will encode to hex
+func (ppc *hexPubkeyConverter) SilentEncode(pkBytes []byte, log core.Logger) string {
```
Comment (iulianpascalau) [inline core/pubkeyConverter/hexPubkeyConverter.go:63]: unit test?

File: core/pubkeyConverter/bech32PubkeyConverter.go (hunk 17 above)
Comment (schimih) [reply core/pubkeyConverter/bech32PubkeyConverter.go:103]: added

File: core/pubkeyConverter/hexPubkeyConverter.go (hunk 18 above)
Comment (schimih) [reply core/pubkeyConverter/hexPubkeyConverter.go:63]: added

**Review by sstanculeanu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]
